File: plugins/extract-ts/extract.ts
// One service in, one fragment out. A fragment, not a catalog: it carries one
// context and one service, names peers it does not own, and is merged with
// everything else before anything validates it.

import { existsSync, readFileSync } from "node:fs";
import { basename, join, relative, resolve } from "node:path";
import type { Catalog, Service } from "../../src/catalog.ts";
import { readAggregates, type Diagnostics } from "./domain.ts";
import { operationOf, readUseCases } from "./operations.ts";
import { readBindings } from "./wiring.ts";
import { readGrpcTransport, readTransport } from "./transport.ts";
import { FlowReader } from "./flows.ts";
import { serviceID, title } from "./ids.ts";

export interface Options {
  context?: string;
  contextName?: string;
  contextSummary?: string;
  classification?: "core" | "supporting" | "generic";
  service?: string;
  serviceName?: string;
  repo?: string;
  store?: string;
  peers?: Record<string, string>;
  events?: Record<string, string>;
  source?: string;
  out?: string;
}

export interface Input {
  root: string;
  commit: string;
  generatedAt: string;
}

export interface Diagnostic {
  severity: "warning" | "error";
  message: string;
  ref?: string;
}

export interface Response {
  files: { name: string; contents: string }[];
  diagnostics: Diagnostic[];
}

class Builder implements Diagnostics {
  files: { name: string; contents: string }[] = [];
  diagnostics: Diagnostic[] = [];
  warn(ref: string, message: string): void {
    this.diagnostics.push({ severity: "warning", message, ref });
  }
}

export function extract(input: Input, opts: Options, cwd = process.cwd()): Response {
  const b = new Builder();
  const root = resolve(cwd, input.root);
  const rel = (abs: string): string => relative(cwd, abs).split("\\").join("/");
  const src = join(root, opts.source ?? "src");

  const context = opts.context || basename(root);
  const service = opts.service || basename(root);
  const svcID = serviceID(context, service);
  const readme = existsSync(join(root, "README.md")) ? readFileSync(join(root, "README.md"), "utf8").trim() : "";

  const aggregates = readAggregates(join(src, "domain"), svcID, rel, b);
  const useCases = readUseCases(join(src, "application"), rel, b);
  const bindings = readBindings(join(src, "di"));
  const transport = readTransport(join(src, "infrastructure", "transport", "http"), rel, b);
  // A service may answer over both, and a service that answers over neither is
  // read the same way with nothing to show for it.
  transport.endpoints.push(...readGrpcTransport(join(src, "infrastructure", "transport", "grpc"), rel, b));

  // Operations belong to the aggregate their use case sits under.
  const exposedBy = new Map<string, string[]>();
  for (const endpoint of transport.endpoints) {
    for (const key of endpoint.useCases) exposedBy.set(key, [...(exposedBy.get(key) ?? []), endpoint.id]);
  }
  for (const uc of useCases) {
    const agg = aggregates.find((a) => a.aggregate.slug === uc.aggregate);
    if (!agg) {
      b.warn(svcID, `${rel(uc.dir)} sits under application/${uc.aggregate}, but there is no matching aggregate under domain`);
      continue;
    }
    const op = operationOf(uc);
    const routes = exposedBy.get(uc.key);
    if (routes) op.exposedBy = [...routes].sort();
    agg.aggregate.operations.push(op);
  }
  for (const agg of aggregates) agg.aggregate.operations.sort((a, c) => a.id.localeCompare(c.id));

  const reader = new FlowReader(
    { context, svcID, service, store: opts.store ?? "", peers: opts.peers ?? {}, events: opts.events ?? {} },
    new Map(useCases.map((u) => [u.key, u])),
    bindings,
    aggregates,
    rel,
    b,
  );
  const flows = [];
  for (const endpoint of transport.endpoints) {
    const flow = reader.endpointFlow(endpoint);
    if (flow) flows.push(flow);
  }
  flows.push(...reader.policyFlows(join(src, "application", "policy")));
  for (const agg of aggregates) {
    for (const [, id] of agg.events) {
      if (!reader.referenced.has(id)) b.warn(id, "no flow reaches this event: nothing this extractor could follow publishes it");
    }
  }

  const svc: Service = {
    id: svcID,
    slug: service,
    name: opts.serviceName || readmeTitle(readme) || title(service),
    repo: opts.repo || packageRepo(root),
    path: rel(root),
    readme,
    provides: [],
    consumes: reader.consumes(),
    aggregates: aggregates.map((a) => a.aggregate),
  };
  if (aggregates.length === 0) b.warn(svcID, "no aggregates found under domain; the fragment describes a service with no model");

  const fragment: Catalog = {
    generatedAt: input.generatedAt,
    commit: input.commit,
    contexts: [
      {
        id: context,
        slug: context,
        name: opts.contextName || title(context),
        summary: opts.contextSummary ?? "",
        ...(opts.classification ? { classification: opts.classification } : {}),
        services: [svc],
      },
    ],
    defs: {},
    flows,
    adrs: [],
  };
  b.files.push({ name: opts.out || "domain.json", contents: `${JSON.stringify(fragment, null, 2)}\n` });
  return { files: b.files, diagnostics: b.diagnostics };
}

function readmeTitle(md: string): string {
  for (const line of md.split("\n")) {
    const t = line.trim();
    if (t.startsWith("# ")) return t.slice(2).trim();
  }
  return "";
}

/** package.json's repository, spelled the way go.mod spells a module: host/owner/name. */
function packageRepo(root: string): string {
  const path = join(root, "package.json");
  if (!existsSync(path)) return "";
  try {
    const pkg = JSON.parse(readFileSync(path, "utf8")) as { repository?: string | { url?: string } };
    const url = typeof pkg.repository === "string" ? pkg.repository : (pkg.repository?.url ?? "");
    return url
      .replace(/^git\+/, "")
      .replace(/^(https?|ssh):\/\//, "")
      .replace(/^git@/, "")
      .replace(":", "/")
      .replace(/\.git$/, "")
      .replace(/\/$/, "");
  } catch {
    return "";
  }
}

File: plugins/extract-rust/src/extract.rs
//! One service in, one fragment out. A fragment, not a catalog: it carries one
//! context and one service, names peers it does not own, and is merged with
//! everything else before anything validates it.

use std::collections::BTreeMap;
use std::fs;
use std::path::{Path, PathBuf};

use crate::catalog::{Catalog, Context, Service};
use crate::domain::read_aggregates;
use crate::flows::{FlowOptions, FlowReader};
use crate::ids::{service_id, title};
use crate::operations::{operation_of, read_use_cases};
use crate::protocol::{Builder, File, Input, Options, Response};
use crate::source::Crate;
use crate::transport::read_transport;
use crate::wiring::read_bindings;

pub fn extract(input: &Input, opts: &Options, cwd: &Path) -> Response {
    let mut b = Builder::default();
    let root = cwd.join(&input.root);
    let cwd_owned = cwd.to_path_buf();
    let rel = move |abs: &Path| -> String { abs.strip_prefix(&cwd_owned).unwrap_or(abs).to_string_lossy().replace('\\', "/") };
    let src = root.join(if opts.source.is_empty() { "src" } else { opts.source.as_str() });

    let base = root.file_name().map(|n| n.to_string_lossy().to_string()).unwrap_or_default();
    let context = if opts.context.is_empty() { base.clone() } else { opts.context.clone() };
    let service = if opts.service.is_empty() { base.clone() } else { opts.service.clone() };
    let svc_id = service_id(&context, &service);
    let readme = fs::read_to_string(root.join("README.md")).map(|s| s.trim().to_string()).unwrap_or_default();

    let krate = Crate::load(&src);
    let mut aggregates = read_aggregates(&krate, &src.join("domain"), &svc_id, &rel, &mut b);
    let use_cases = read_use_cases(&krate, &src.join("application"), &rel, &mut b);
    let bindings = read_bindings(&krate);
    let endpoints = read_transport(&krate, &src.join("infrastructure").join("transport").join("grpc"), &rel, &mut b);

    // Operations belong to the aggregate their use case sits under.
    let mut exposed_by: BTreeMap<String, Vec<String>> = BTreeMap::new();
    for endpoint in &endpoints {
        for key in &endpoint.use_cases {
            exposed_by.entry(key.clone()).or_default().push(endpoint.id.clone());
        }
    }
    for uc in &use_cases {
        let Some(agg) = aggregates.iter_mut().find(|a| a.aggregate.slug == uc.aggregate) else {
            b.warn(
                &svc_id,
                format!(
                    "{} sits under application/{}, but there is no matching aggregate under domain",
                    rel(&uc.dir),
                    uc.aggregate
                ),
            );
            continue;
        };
        let mut op = operation_of(uc);
        if let Some(routes) = exposed_by.get(&uc.key) {
            let mut routes = routes.clone();
            routes.sort();
            op.exposed_by = Some(routes);
        }
        agg.aggregate.operations.push(op);
    }
    for agg in &mut aggregates {
        agg.aggregate.operations.sort_by(|a, c| a.id.cmp(&c.id));
    }

    let mut flows = Vec::new();
    let consumes;
    let referenced;
    {
        let mut reader = FlowReader::new(
            FlowOptions {
                context: context.clone(),
                svc_id: svc_id.clone(),
                service: service.clone(),
                store: opts.store.clone(),
                peers: opts.peers.clone(),
                events: opts.events.clone(),
            },
            &krate,
            &use_cases,
            &bindings,
            &aggregates,
            &rel,
            &mut b,
        );
        for endpoint in &endpoints {
            flows.push(reader.endpoint_flow(endpoint));
        }
        flows.extend(reader.policy_flows(&src.join("application").join("policy")));
        consumes = reader.consumes();
        referenced = reader.referenced.clone();
    }
    for agg in &aggregates {
        for id in agg.events.values() {
            if !referenced.contains(id) {
                b.warn(id, "no flow reaches this event: nothing this extractor could follow publishes it");
            }
        }
    }

    let name = if !opts.service_name.is_empty() {
        opts.service_name.clone()
    } else {
        readme_title(&readme).unwrap_or_else(|| title(&service))
    };
    let svc = Service {
        id: svc_id.clone(),
        slug: service,
        name,
        repo: if opts.repo.is_empty() { cargo_repo(&root) } else { opts.repo.clone() },
        path: rel(&root),
        readme,
        provides: vec![],
        consumes,
        aggregates: aggregates.into_iter().map(|a| a.aggregate).collect(),
    };
    if svc.aggregates.is_empty() {
        b.warn(&svc_id, "no aggregates found under domain; the fragment describes a service with no model");
    }

    let fragment = Catalog {
        generated_at: input.generated_at.clone(),
        commit: input.commit.clone(),
        contexts: vec![Context {
            id: context.clone(),
            slug: context.clone(),
            name: if opts.context_name.is_empty() {
                title(&context)
            } else {
                opts.context_name.clone()
            },
            summary: opts.context_summary.clone().unwrap_or_default(),
            classification: opts.classification.clone(),
            services: vec![svc],
        }],
        defs: serde_json::Map::new(),
        flows,
        adrs: vec![],
    };
    let contents = serde_json::to_string_pretty(&fragment).unwrap_or_default() + "\n";
    b.files.push(File {
        name: if opts.out.is_empty() { "domain.json".into() } else { opts.out.clone() },
        contents,
    });
    Response {
        files: b.files,
        warnings: b.warnings,
        describe: None,
    }
}

fn readme_title(md: &str) -> Option<String> {
    md.lines().map(str::trim).find_map(|l| l.strip_prefix("# ").map(|t| t.trim().to_string()))
}

/// Cargo.toml's repository, spelled the way go.mod spells a module: host/owner/name.
fn cargo_repo(root: &Path) -> String {
    let Ok(text) = fs::read_to_string(root.join("Cargo.toml")) else {
        return String::new();
    };
    let url = text
        .lines()
        .map(str::trim)
        .find_map(|l| {
            l.strip_prefix("repository")
                .and_then(|r| r.trim().strip_prefix('='))
                .map(|v| v.trim().trim_matches('"').to_string())
        })
        .unwrap_or_default();
    url.trim_start_matches("git+")
        .trim_start_matches("https://")
        .trim_start_matches("http://")
        .trim_start_matches("ssh://")
        .trim_start_matches("git@")
        .replacen(':', "/", if url.starts_with("git@") { 1 } else { 0 })
        .trim_end_matches(".git")
        .trim_end_matches('/')
        .to_string()
}

pub fn _unused(_: PathBuf) {}
